### PR TITLE
Feature/sim kv fault injection

### PR DESCRIPTION
**Pull request by ahrav**
## Summary by CodeRabbit

* **New Features**
  * Deterministic ID/hash behavior and clearer error diagnostics for more predictable outcomes
  * Runtime fault‑injection profiles in the simulator to exercise transient failures, uncertain commits, lease races, and retry‑exhaustion bursts

* **Tests**
  * Expanded deterministic, property, and integration tests covering boundary conditions, retry exhaustion, lease TTL races, uncertain commits, and fingerprinting edge cases

* **Documentation**
  * Coordination testing docs updated to highlight the deterministic simulator and new fault scenarios

### CODE REVIEW

File: crates/gossip-coordination-etcd/src/sim_etcd_kv.rs [hunk 1]
```diff
@@ -366,6 +415,96 @@ impl SimulatedEtcdKV {
...
+        true
+    }
+
+    fn maybe_arm_retry_exhaustion(&mut self, request: &PbTxnRequest, fingerprint: u64) -> bool {
+        debug_assert!(
+            self.fault_config.retry_exhaustion_ppm == 0
+                || self.fault_config.retry_exhaustion_attempts > 0,
+            "retry_exhaustion_attempts is 0 but retry_exhaustion_ppm is nonzero; \
+             the fault will never fire"
+        );
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/gossip-coordination-etcd/src/sim_etcd_kv.rs]: Good catch — confirmed. Wrote a test targeting the `ppm > 0, attempts == 0` misconfiguration and verified the `debug_assert!` is indeed unreachable: `retry_exhaustion_possible` is `false` when `attempts == 0`, so `maybe_arm_retry_exhaustion` is never called.

**Fix:** Moved the `debug_assert!` to `with_fault_config()` and `set_fault_config()` where it fires at configuration time. Removed the unreachable one from `maybe_arm_retry_exhaustion`. Also fixed `arb_fault_config()` in the proptests to generate valid configs.

File: crates/gossip-coordination-etcd/src/sim_etcd_kv.rs [hunk 2]
```diff
@@ -366,6 +415,96 @@ impl SimulatedEtcdKV {
...
+
+        if request.compare.is_empty()
+            || request.success.is_empty()
+            || self.fault_config.retry_exhaustion_attempts == 0
+            || self
+                .recent_retry_exhaustion
+                .is_some_and(|recent| recent == fingerprint)
+            || !self.fault_config.should_retry_exhaust(&mut self.rng)
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/gossip-coordination-etcd/src/sim_etcd_kv.rs]: The `recent_retry_exhaustion` cooldown is intentional — it prevents back-to-back retry-exhaustion bursts on the same logical CAS operation. Without it, a successful burst (where the coordinator eventually retries through) could immediately trigger another burst on the very next attempt of the same operation, making the fault disproportionately punishing.

In practice, the coordinator interleaves operations across multiple shards, so a different fingerprint arrives between repeated operations on the same shard, naturally clearing the cooldown. For a single-shard homogeneous workload the fault fires once and then clears on the next different operation — this matches the real-world scenario where sustained contention on one key is broken by intervening traffic.

No change for now, but worth revisiting if simulation results show insufficient coverage on homogeneous workloads.

File: crates/gossip-coordination-etcd/src/sim_etcd_kv.rs [hunk 3]
```diff
@@ -244,6 +267,22 @@ impl SimulatedEtcdKV {
             });
         }
 
+        let retry_exhaustion_possible = self.active_retry_exhaustion.is_some()
+            || self.recent_retry_exhaustion.is_some()
+            || (self.fault_config.retry_exhaustion_ppm > 0
+                && self.fault_config.retry_exhaustion_attempts > 0);
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/gossip-coordination-etcd/src/sim_etcd_kv.rs]: Confirmed — this is the same underlying issue as the unreachable `debug_assert!` flagged by greptile. Setting `retry_exhaustion_ppm > 0` with the default `attempts == 0` silently disables the fault.

**Fix:** Added a `debug_assert!` to both `with_fault_config()` and `set_fault_config()` that fires at configuration time when `ppm > 0 && attempts == 0`, catching the misconfiguration early instead of silently doing nothing.

**Comment by ahrav** [comment]
### Responses to Greptile "Outside Diff" Findings

**1. Lease TTL race in `lease_keep_alive_once` — CONFIRMED & FIXED**

Confirmed the infinite retry loop. Wrote a test positioning the clock 1 tick before lease expiry with 100% TTL race — `lease_keep_alive_once` advances the clock, expires the lease, then returns `LeaseNotFound` which `map_etcd_err` maps to `Transient`. The coordinator retries forever against a permanently gone lease.

**Fix:** Removed the `maybe_inject_lease_ttl_race()` call from `lease_keep_alive_once`. The TTL race is only meaningful in `txn()` where it tests CAS boundary behavior with lease-guarded keys. In `lease_keep_alive_once`, it creates an unrecoverable condition. Updated the existing test (`lease_ttl_race_expires_lease_during_keep_alive` → `keep_alive_ignores_ttl_race_fault`) to verify the corrected behavior.

**2. `UncertainCommit` ordering invariant — defensive comment added**

Agreed this is safe but fragile. Added an inline comment at the `UncertainCommit` return site documenting why the retry-exhaustion state machine is safe to leave unchanged: on retry, the compare predicates fail (revision changed by the committed write), so we exit before `maybe_arm_retry_exhaustion` can fire.
